text/2018-08-29-destroy-range.md

```diff
@@ -0,0 +1,50 @@
+# Summary
+
+Support RPC `DestroyRange`. This call is on the whole TiKV rather than a certain region. When it is invoked, TiKV will use `delete_files_in_range` to quickly free a large amount of space and delete all keys in the range. Raft layer will be bypassed, and the range can cover across multiple regions. The invoker should promise that after invoking `DestroyRange`, the range will never be accessed again. That is to say, the range is permanently scrapped.
+
+This interface is only designed for TiDB.
+
+# Motivation
+
+Currently, when TiDB deletes a range, it will invoke `DeleteRange` on all affected regions respectively. Though TiKV will call `delete_files_in_range` to quickly free up the space for each region, the SST files that span over multiple regions can't be removed immediately. As a result, the disk space can't be recycled in time. Another problem is, the process to delete all regions respectively is too slow.
+
+This issue has troubled some of our users. It should be fixed as soon as possible.
+
+# Detailed design
+
+1. Add protobuf for `DestroyRange` to project kvrpcpb.
+    ```protobuf
+    message DestroyRangeRequest {
+        bytes start_key = 1;
+        bytes end_key = 2;
+    }
+    ```
+    It doesn't contains a context to indicate which region to work on.
+
+2. When TiKV receives `DestroyRangeRequest`, it immediately run `delete_files_in_range` on whole range on RocksDB, bypassing the Raft layer. Then it will scan and delete all remaining keys in the range. Here is the pseudo code of the procedure:
+    > TODO: Describe more detailed implementation here
+
+    The story of TiKV is ended here. But to show why this thing is useful, let's continue to see what we will do on TiDB:
+
+3. In TiDB, there are two system tables about DeleteRanges (the second step of GC): `gc_delete_range` and `gc_delete_range_done`. Currently, ranges to be deleted are recorded in `gc_delete_range`. When we finished deleting a range, we move the record from `gc_delete_range` to `gc_delete_range_done`. So in the new implementation:
```

Comment:
In TiDB, when you execute `TRUNCATE/DROP TABLE/INDEX`, the meta data of the related table will be modified at first, and then the range covered by the truncated/dropped table/index will be recorded in the system table `gc_delete_range` with timestamp when the `TRUNCATE/DROP` query is executed. After the recorded range has passed GC lifetime, TiDB gc_worker will ...